File: software/examples/c/example_modbus_slave.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_rs485.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your RS232 Bricklet

RS485 rs485;

// Callback function for write single register request callback
void cb_modbus_slave_write_single_register_request(uint8_t request_id,
                                                   uint32_t register_address,
                                                   uint16_t register_value,
                                                   void *user_data) {
    printf("Request ID = %d\n", request_id);
    printf("Register Address = %d\n", register_address);
    printf("Register Value = %d\n", register_value);

    // Here we assume valid writable register address is 42
    if(register_address != 42) {
      printf("Requested invalid register address\n");
      rs485_modbus_slave_report_exception(&rs485, request_id, RS485_EXCEPTION_CODE_ILLEGAL_DATA_ADDRESS);
    }
    else {
      printf("Request OK\n");
      rs485_modbus_slave_answer_write_single_register_request(&rs485, request_id);
    }
}

int main(void) {
    // Create IP connection
    IPConnection ipcon;
    ipcon_create(&ipcon);

    // Create device object
    rs485_create(&rs485, UID, &ipcon);

    // Connect to brickd
    if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
        fprintf(stderr, "Could not connect\n");

        return 1;
    }
    // Don't use device before ipcon is connected

    // Set operating mode of the Bricklet
    rs485_set_mode(&rs485, RS485_MODE_MODBUS_SLAVE_RTU);

    /*
     * Modbus specific configuration
     *
     * Slave mode address = 1 (Unused in master mode)
     * Master mode request timeout = 1000ms (Unused in slave mode)
     */
    rs485_set_modbus_configuration(&rs485, 1, 1000);

    // Register write single register request callback
    rs485_register_callback(&rs485,
                            RS485_CALLBACK_MODBUS_SLAVE_WRITE_SINGLE_REGISTER_REQUEST,
                            (void *)cb_modbus_slave_write_single_register_request,
                            NULL);

    printf("Press key to exit\n");
    getchar();

    rs485_destroy(&rs485);
    ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally

    return 0;
}

File: software/examples/php/ExampleModbusSlave.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletRS485.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletRS485;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your RS485 Bricklet

// Callback function for Modbus slave write single register request callback
function cb_modbusSlaveWriteSingleRegisterRequest($request_id, $register_address,
                                                  $register_value, $user_data)
{
    $rs485 = $user_data;

    echo "Request ID: $request_id\n";
    echo "Register Address: $register_address\n";
    echo "Register Value: $register_value\n";

    if ($register_address != 42) {
        echo "Requested invalid register address\n";
        $rs485->modbusSlaveReportException($request_id, BrickletRS485::EXCEPTION_CODE_ILLEGAL_DATA_ADDRESS);
    } else {
        $rs485->modbusSlaveAnswerWriteSingleRegisterRequest($request_id);
    }
}

$ipcon = new IPConnection(); // Create IP connection
$rs485 = new BrickletRS485(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Set operating mode to Modbus RTU slave
$rs485->setMode(BrickletRS485::MODE_MODBUS_SLAVE_RTU);

// Modbus specific configuration:
// - slave address = 17
// - master request timeout = 0ms (unused in slave mode)
$rs485->setModbusConfiguration(17, 0);

// Register Modbus slave write single register request callback to function
// cb_modbusSlaveWriteSingleRegisterRequest
$rs485->registerCallback(BrickletRS485::CALLBACK_MODBUS_SLAVE_WRITE_SINGLE_REGISTER_REQUEST,
                         'cb_modbusSlaveWriteSingleRegisterRequest', $rs485);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
